### PR TITLE
Refactor the training algorithm infrastructure; add first-class LTFB

### PR DESCRIPTION
**Pull request by benson31**
There's a lot here; please start reviewing at your earliest convenience. The remaining work is either superficial or Python-related.

TODO:

- [x] Add new "unit_test" for LTFB training algo to mirror the LTFB callback test.
- [x] Port the recent LTFB callback changes to `SendRecvWeights` into the corresponding `ExchangeStrategy` in this PR (I did the initial migration from the callback before that refactor happened).
- [x] Fix the `SendRecvWeights` and `CheckpointFile` `ExchangeStrategy`s so they don't return `nullptr`.
- [x] Add catch tests to verify protobuf/factory robustness, re `google::protobuf::Any`.
- [x] Finish and publish the PFE changes for this work.

Future work, most likely reserved for future PRs:
- Publish user and (downstream) developer documentation.
- Refactor the SGD training algorithm.
- Clean up the output/factor into a unified logging framework.

### CODE REVIEW

File: python/lbann/core/training_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+        self.name = name
+
+    def export_proto(self):
+        """Get a protobuf representation of this object."""
+
+        algo = AlgoProto.TrainingAlgorithm()
+        algo.name = self.name
+        algo.parameters.Pack(self.do_export_proto())
+        return algo
+
```
Comment (timmoon10) [inline python/lbann/core/training_algorithm.py:41]:
```suggestion

    def do_export_proto(self):
        """Get a protobuf representation of this object.
        
        Must be implemented in derived classes.  
        """
        raise NotImplementedError

```

**Review by timmoon10 — APPROVED** [review]
LGTM

File: include/lbann/execution_algorithms/ltfb/random_pairwise_exchange.hpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+   *  @param[in] weights_names Names of weights to exchange. If empty,
+   *                           then all weights are exchanged.
+   *  @param[in] exchange_hyperparameters Exchange optimizer
+   *                                      hyperparameters.
+   */
+  SendRecvWeights(std::set<std::string> const& weights_names,
+                  bool exchange_hyperparameters);
+
+  /** @brief Construct from weights names
+   *  @param[in] weights_names Names of weights to exchange. If empty,
```
Comment (samadejacobs) [inline include/lbann/execution_algorithms/ltfb/random_pairwise_exchange.hpp:203]: Part of unit test (if any) should cover this (that is sendrcv of selected weights)

File: python/lbann/core/training_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+                     keep_weights_names: list[str] = [],
+                     exchange_hyperparameters: bool = False,
+                     checkpoint_dir: str = None):
+            """Construct a new exchange strategy.
+
+            Args:
+                strategy:
+                  Which strategy to use (default: "checkpoint_binary").
+                keep_weights_names:
+                  A list of weights names that should not be exchanged.
```
[the remark this answers is not in the document]
Comment (samadejacobs) [reply python/lbann/core/training_algorithm.py:263]: Seems we are reversing how it is currently done where the API exposes a list of weights names to be exchanged. This "reverse implementation" is fine as long as it is correct. We make a note to make sure this change is reflected in existing applications (e.g., ICF JAG and ATOM)

File: src/proto/training_algorithm.proto [hunk 4]
```diff
@@ -26,9 +26,73 @@
...
+}// message LTFB
+
+// The classic LTFB algorithm. Implements MetaLearningStrategy.
+message RandomPairwiseExchange {
+  enum MetricStrategy {
+    LOWER_IS_BETTER = 0;
+    HIGHER_IS_BETTER = 1;
+  }
+
+  string metric_name = 1;
```
Comment (samadejacobs) [inline src/proto/training_algorithm.proto:74]: Immediate extension of this will be support for multiple evaluation metrics, each metric with its own metric strategy.

**Review by samadejacobs — COMMENTED** [review]
Generally LGTM, few comments.

File: include/lbann/execution_algorithms/ltfb/termination_criteria.hpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+/** @class TerminationCriteria
+ *  @brief The stopping criteria for an LTFB-type algorithm
+ *
+ *  An object here needs to manage
+ */
+class TerminationCriteria final : public lbann::termination_criteria
+{
+public:
+  TerminationCriteria(size_t max_num_steps)
```
[the remark this answers is not in the document]
Comment (benson31) [reply include/lbann/execution_algorithms/ltfb/termination_criteria.hpp:43]: `max_num_steps` is specified in the Python front-end.

A step of the `LTFB` training algorithm is defined as one round of training (which has its own stopping criteria!) followed by one call to the metalearning strategy. Thus, each trainer must specify the same value. It will handle a user's failure to do so by hanging some, possibly all, of the ranks (probably all). I can update this to assert that all trainers do the same number of steps.

File: include/lbann/execution_algorithms/training_algorithm.hpp [hunk 6]
```diff
@@ -97,8 +98,14 @@ class training_algorithm
   ///@{
   /** @brief A string identifying the type of the object. */
   virtual std::string get_type() const = 0;
+
   /** @brief A user-defined string identifying the algorithm object. */
   std::string const& get_name() const noexcept;
+
+  /** @brief Get the current termination criteria for the algorithm. */
+  // virtual termination_criteria& get_termination_criteria() = 0;
```
[the remark this answers is not in the document]
Comment (benson31) [reply include/lbann/execution_algorithms/training_algorithm.hpp]: They were commented out and have been deleted.

File: include/lbann/execution_contexts/sgd_execution_context.hpp [hunk 7]
```diff
@@ -35,8 +35,16 @@ namespace lbann {
...
+                           size_t max_num_epochs)
+    : termination_criteria{max_num_batches},
+      m_max_epochs{max_num_epochs}
+  {
+  }
   ~sgd_termination_criteria() = default;
-  size_t num_epochs;
+  size_t max_num_epochs() const noexcept { return m_max_epochs; }
+private:
+  size_t m_max_epochs;
```
[the remark this answers is not in the document]
Comment (benson31) [reply include/lbann/execution_contexts/sgd_execution_context.hpp:47]: `sgd_execution_context` counts both. The step (batch) count is in the base `execution_context` class, since most of our training algorithms are expected to be iterative.

I strongly dislike the "epoch concept" and would love to purge it from the code.

File: include/lbann/utils/make_abstract.hpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+#include <google/protobuf/message.h>
+#include <type_traits>
+
+namespace lbann {
+/** @brief Template that always has a 'value' field that evaluates to
+ *         'false'.
+ *
+ *  This is mostly for static_asserts so the literal "false" isn't in
+ *  the "expression" argument. Some compilers also give
```
[the remark this answers is not in the document]
Comment (benson31) [reply include/lbann/utils/make_abstract.hpp:37]: Ah yes, there was a compiler error as I was writing it. Incomplete output. 😛

File: python/lbann/core/training_algorithm.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+        return params
+
+class RandomPairwiseExchange(MetaLearningStrategy):
+    """The classic LTFB pairwise tourament metalearning strategy.
+
+    This metalearning strategy is the original algorithm used for
+    LTFB. After each local training step, all trainers in the LBANN
+    environment pair off and have a "tournament" (if the number of
+    trainers is odd, one competes with itself). The tournament is a
+    simple comparison of a metric value that's read from the
```
[the remark this answers is not in the document]
Comment (benson31) [reply python/lbann/core/training_algorithm.py:193]: Arguably that will represent a new strategy. That's part of the point of this refactor/new direction -- the strategy pattern should isolate these decisions into smaller units that are (hopefully) easy to write and register with the factory. When we have a new idea, we just write a new strategy instead of adding a bunch of spaghetti code and confusing switches to shim it into the existing code.

File: python/lbann/core/training_algorithm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (benson31) [reply python/lbann/core/training_algorithm.py:263]: The docstring is bad and/or I didn't read the existing code closely enough. The weights specified here are, in fact, the weights that *WILL* be exchanged. Sorry for the confusion; will fix.

File: include/lbann/utils/make_abstract.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (benson31) [reply include/lbann/utils/make_abstract.hpp]: Fixed.

File: python/lbann/core/training_algorithm.py [hunk 10]
```diff
@@ -0,0 +1,322 @@
+"""LBANN Training Algorithms.
+
+These are methods for training neural network models in LBANN. If a
+user does not explicitly specify a training algorithm in their
+lbann.Trainer object, the default is equivalent to:
+
+BatchedDescent('Some name', epoch_count=1)
```
[the remark this answers is not in the document]
Comment (benson31) [reply python/lbann/core/training_algorithm.py]: Addressed.

File: python/lbann/core/training_algorithm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (benson31) [reply python/lbann/core/training_algorithm.py]: Fixed.

File: include/lbann/execution_contexts/sgd_execution_context.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (bvanessen) [reply include/lbann/execution_contexts/sgd_execution_context.hpp]: I think that this kind of purge is better handled when we finish reworking the data readers / data source / data coordinator.

**Review by mrwyattii — APPROVED** [review]
LGTM

**Review by bvanessen — APPROVED** [review]
